controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
-			return newHomeCluster, nil
-		case d.HomeCluster:
-			return newHomeCluster, errorswrapper.Wrap(ErrSameHomeCluster, newHomeCluster)
-		case d.PrevHomeCluster:
-			return newHomeCluster,
-				fmt.Errorf("miconfiguration detected on failover! (n:%s,p:%s)", newHomeCluster, d.PrevHomeCluster)
+func (a *AVRInstance) findManifestWork(mwName, homeCluster string) (*ocmworkv1.ManifestWork, error) {
+	if homeCluster != "" {
+		mw := &ocmworkv1.ManifestWork{}
+
+		err := a.reconciler.Get(a.ctx, types.NamespacedName{Name: mwName, Namespace: homeCluster}, mw)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return nil, fmt.Errorf("MW not found (%w)", err)
+			}
+
+			return nil, fmt.Errorf("failed to retrieve manifestwork (%w)", err)
 		}
+
+		return mw, nil
+	}
+
+	return nil, nil
+}
+
+func (a *AVRInstance) isVRGPrimary(mw *ocmworkv1.ManifestWork) (bool, error) {
+	vrgClientManifest := &mw.Spec.Workload.Manifests[0]
+	vrg := &rmn.VolumeReplicationGroup{}
+
+	err := yaml.Unmarshal(vrgClientManifest.RawExtension.Raw, &vrg)
+	if err != nil {
+		return false, errorswrapper.Wrap(err, fmt.Sprintf("unable to get vrg from ManifestWork %s", mw.Name))
 	}
 
-	return newHomeCluster, fmt.Errorf("unknown error (n:%s,avrStatus:%v)", newHomeCluster, a.instance.Status)
+	return (vrg.Spec.ReplicationState == volrep.Primary), nil
 }
 
-func (a *AVRInstance) validateFailback(
-	subscription *subv1.Subscription, newHomeCluster string) (string, error) {
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		switch newHomeCluster {
-		case d.PrevHomeCluster:
-			return newHomeCluster, nil
-		case d.HomeCluster:
-			return newHomeCluster, errorswrapper.Wrap(ErrSameHomeCluster, newHomeCluster)
-		case d.PeerCluster:
-			return "", fmt.Errorf("miconfiguration detected on failover! (n:%s,p:%s)", newHomeCluster, d.PrevHomeCluster)
+func (a *AVRInstance) ensureCleanup() error {
+	a.log.Info("ensure cleanup ", "cluster", a.instance.Status.ClusterToClean)
+
+	if a.instance.Status.ClusterToClean != "" {
+		_, err := a.ensureOldVRGMWOnClusterDeleted(a.instance.Status.ClusterToClean)
+		if err != nil {
+			return err
 		}
 	}
 
-	return "", fmt.Errorf("unknown error (n:%s,avrStatus:%v)", newHomeCluster, a.instance.Status)
+	return nil
 }
 
-func (a *AVRInstance) selectPlacementDecision(
-	subscription *subv1.Subscription) (string, string, error) {
-	a.log.Info("Selecting placement decisions for subscription", "name", subscription.Name)
+func (a *AVRInstance) ensureOldVRGMWOnClusterDeleted(clusterName string) (bool, error) {
+	a.log.Info("Ensuring that previous cluster VRG MW is deleted for ", "cluster", clusterName)
+
+	const done = true
+
+	mwName := BuildManifestWorkName(a.instance.Name, a.instance.Namespace, MWTypeVRG)
+	mw := &ocmworkv1.ManifestWork{}
 
-	// get the placement rule fo this subscription
-	placementRule, err := a.getPlacementRule(subscription)
+	err := a.reconciler.Get(a.ctx, types.NamespacedName{Name: mwName, Namespace: clusterName}, mw)
```

Comment:
(question/future) What is the MW reconcile assurance? Is it deleted only when MW on ManagedCluster is deleted? Or, is it deleted ASAP on hub, and ManagedCluster catches up to the fact? If the former, then just checking if MW is gone on the hub is a good enough check for our purposes.

Reply:
I Don't know.  We'll test it later.